rfcs/0006-configuration-constants.md

```diff
@@ -0,0 +1,99 @@
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+When a Sway program is compiled, a set of artifacts are produced. Included in these artifacts are a JSON ABI descriptor, a configuration-time-constant descriptor, and the actual bytecode. The object we are discussing here is the configuration-time-constant descriptor, which describes the specific offset within the bytecode, in terms of bytes, to the data section entry for a particular configuration variable.
+
+The configuration variables for a program can be identified by the `configurable` keyword within the Sway language. This denotes a constant within the program that has no known value at compile time. The value will be provided at configuration time by the SDK or some other consumer of the configuration-time-constant descriptor. As an example:
+
+```rust
+// main.sw
+script;
+
+configurable CONTRACT_ADDRESS: b256;
+
+fn main() { ... }
+```
+```
+// configurable-constants.json
+
+{
+    "CONTRACT_ADDRESS": {
+        "type": "b256",
+        "offset": "12345"
+    }
+}
+```
+
+If you are not using the SDK, it will be possible to define these values either by passing a flag to `forc` or including them in the `Forc.toml` manifest file.
+
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The `type` field of the JSON file should describe the type of the configurable in the same manner as types are described in the JSON ABI file.
+
+No two configurables in a program should have the same name to prevent collisions.
+
+The range from the offset to the offset plus the size of the type will represent bytes within the bytecode that will be overwritten by the SDK, which will write values according to the ABI encoder's memory layout, which is consistent with Sway's memory layout.
+
+The change is not breaking as it is entirely new functionality.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+One major drawback is the additional cognitive complexity this adds to the compilation process. If any bug is introduced or the overwriting of the bytecode goes awry in any way, it will result in confusing and inconsistent undefined behavior. We will need to introduce sufficient checks to ensure that an end user would never encounter this situation.
+
+This increases our reliance on the correctness of the ABI encoder in the SDK, and any version mismatches or bugs in the encoder will result in similarly undefined and unpredictable behavior.
+
+This change also further couples the language to the SDK, further minimizing the use case of writing Sway without the SDK. This can be alleviated by introducing mechanisms to `forc` for defining these values manually, as mentioned in the guide-level explanation.
```

Comment:
This also potentially means that we are going back to the old issue of circular deps between `fuels-rs` and `sway`?

Reply:
Since we would rely  on the abi decoder/encoder for this, we hopefully would avoid having to import `sway`. In fact, it is a goal of this effort to avoid having to import `sway`, so if it isn't possible, we will have to refactor.